runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/ExecutionStateTracker.java

```diff
@@ -97,6 +98,17 @@ public void onActivate(boolean pushing) {}
      */
     public abstract void reportLull(Thread trackedThread, long millis);
 
+    /**
+     * Called when a lull has been detected since the start of bundle processing. This indicates
+     * that more than {@link #BUNDLE_LULL_REPORT_MS} has been spent processing the same bundle (ie
+     * between startBundle/finishBundle).
+     *
+     * @param trackedThread The execution thread that is in a lull.
+     * @param millis The milliseconds since the state was most recently entered.
+     */
+    public abstract void reportBundleLull(
```

Comment:
This would trigger a bigger change because the logging is happening at DataflowExecutionState which is an implementation of ExecutionState.

Reply:
I think that is worthwhile to avoid having this rely on the step context when it should be logged regardless of if there is a step context or not.

You can extract whatever shared logic to print the thread stacks static method in ExecutionStateTracker to share it.